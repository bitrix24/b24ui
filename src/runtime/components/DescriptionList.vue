<script lang="ts">
import type { AppConfig } from '@nuxt/schema'
import theme from '#build/b24ui/description-list'
import type { AvatarProps, ButtonProps, IconComponent } from '../types'
import type { DynamicSlots, ComponentConfig } from '../types/utils'

type DescriptionList = ComponentConfig<typeof theme, AppConfig, 'descriptionList'>

export interface DescriptionListItem {
  label?: string
  /**
   * Display an icon on the left side.
   * @IconComponent
   */
  icon?: IconComponent
  avatar?: AvatarProps
  slot?: string
  description?: string
  /**
   * The orientation between the content and the actions.
   * @defaultValue 'vertical'
   */
  orientation?: DescriptionList['variants']['orientation']
  /**
   * Display a list of actions:
   * - under the description when orientation is `vertical`
   * - next to the description when orientation is `horizontal`
   * `{ size: 'xs' }`{lang="ts"}
   */
  actions?: ButtonProps[]
  class?: any
  b24ui?: Partial<typeof descriptionList.slots>
  [key: string]: any
}

export interface DescriptionListProps<T extends DescriptionListItem = DescriptionListItem> {
  legend?: string
  text?: string
  /**
   * The key used to get the label from the item.
   * @defaultValue 'label'
   */
  labelKey?: string
  /**
   * The key used to get the description from the item.
   * @defaultValue 'description'
   */
  descriptionKey?: string
  items?: T[]
  /**
   * @defaultValue 'md'
   */
  size?: DescriptionList['variants']['size']
  class?: any
  b24ui?: DescriptionList['slots']
}

type SlotProps<T extends DescriptionListItem> = (props: { item: T, index: number }) => any

export type DescriptionListSlots<T extends DescriptionListItem = DescriptionListItem> = {
  legend(props?: {}): any
  text(props?: {}): any
  leading: SlotProps<T>
  label: SlotProps<T>
  description: SlotProps<T>
  actions: SlotProps<T>
  content: SlotProps<T>
  footer(props?: { b24ui: any }): any
} & DynamicSlots<T, undefined, { index: number }>
</script>

<script setup lang="ts" generic="T extends DescriptionListItem">
import { computed } from 'vue'
import { useAppConfig } from '#imports'
import { get } from '../utils'
import { tv } from '../utils/tv'
import B24Avatar from './Avatar.vue'
import B24Button from './Button.vue'

const props = withDefaults(defineProps<DescriptionListProps<T>>(), {
  labelKey: 'label',
  descriptionKey: 'description'
})
const slots = defineSlots<DescriptionListSlots<T>>()

const appConfig = useAppConfig() as DescriptionList['AppConfig']

const b24ui = computed(() => tv({ extend: tv(theme), ...(appConfig.b24ui?.descriptionList || {}) })({
  size: props.size
}))

function normalizeItem(item: any) {
  if (item === null) {
    return {
      label: undefined,
      description: undefined,
      orientation: undefined
    }
  }

  const label = get(item, props.labelKey as string)
  const description = get(item, props.descriptionKey as string)
  const orientation = item?.orientation || 'vertical'

  return {
    ...item,
    label,
    description,
    orientation
  }
}

const normalizedItems = computed(() => {
  if (!props.items) {
    return []
  }

  return props.items.map(normalizeItem)
})
</script>

<template>
  <div :class="b24ui.root({ class: [props.class, props.b24ui?.root] })">
    <h2 v-if="legend || !!slots.legend" :class="b24ui.legend({ class: props.b24ui?.legend })">
      <slot name="legend">
        {{ legend }}
      </slot>
    </h2>
    <p v-if="text || !!slots.text" :class="b24ui.text({ class: props.b24ui?.text })">
      <slot name="text">
        {{ text }}
      </slot>
    </p>

    <dl :class="b24ui.container({ class: props.b24ui?.container })">
      <template
        v-for="(item, index) in normalizedItems"
        :key="index"
      >
        <slot
          :name="((item.slot || 'content') as keyof DescriptionListSlots<T>)"
          :item="(item as Extract<T, { slot: string; }>)"
          :index="index"
        >
          <dt
            :class="b24ui.labelWrapper({
              class: [
                props.b24ui?.labelWrapper,
                item?.b24ui?.labelWrapper
              ]
            })"
          >
            <slot name="leading" :item="item" :index="index">
              <Component
                :is="item.icon"
                v-if="item.icon"
                :class="b24ui.icon({
                  class: [
                    props.b24ui?.icon,
                    item?.b24ui?.icon
                  ]
                })"
              />
              <B24Avatar
                v-else-if="item.avatar"
                :size="((props.b24ui?.avatarSize || b24ui.avatarSize()) as AvatarProps['size'])"
                v-bind="item.avatar"
                :class="b24ui.avatar({
                  class: [
                    props.b24ui?.avatar,
                    item?.b24ui?.avatar
                  ]
                })"
              />
            </slot>
            <span
              :class="b24ui.label({
                class: [
                  item?.class,
                  props.b24ui?.label,
                  item?.b24ui?.label
                ]
              })"
            >
              <slot name="label" :item="item" :index="index">
                {{ item.label }}
              </slot>
            </span>
          </dt>
          <dd
            :data-orientation="item.orientation"
            :class="b24ui.descriptionWrapper({
              class: [
                props.b24ui?.descriptionWrapper,
                item?.b24ui?.descriptionWrapper
              ],
              orientation: item.orientation
            })"
          >
            <span
              :class="b24ui.description({
                class: [
                  item?.class,
                  props.b24ui?.description,
                  item?.b24ui?.description
                ],
                orientation: item.orientation
              })"
            >
              <slot name="description" :item="item" :index="index">
                {{ item.description }}
              </slot>
            </span>
            <span
              v-if="item.actions?.length || !!slots.actions"
              :class="b24ui.actions({
                class: [
                  props.b24ui?.actions,
                  item?.b24ui?.actions
                ],
                orientation: item.orientation
              })"
            >
              <slot name="actions" :item="item" :index="index">
                <B24Button
                  v-for="(action, indexActions) in item.actions"
                  :key="indexActions"
                  size="xs"
                  v-bind="action"
                />
              </slot>
            </span>
          </dd>
        </slot>
      </template>
    </dl>
    <div
      v-if="!!slots.footer"
      :class="b24ui.footer({ class: props.b24ui?.footer })"
    >
      <slot name="footer" :b24ui="b24ui" />
    </div>
  </div>
</template>
