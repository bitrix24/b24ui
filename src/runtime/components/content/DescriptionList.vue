<script lang="ts">
import type { VariantProps } from 'tailwind-variants'
import type { AppConfig } from '@nuxt/schema'
import _appConfig from '#build/app.config'
import theme from '#build/b24ui/content/description-list'
import { tv } from '../../utils/tv'
import type { AvatarProps, ButtonProps, IconComponent } from '../../types'
import type { DynamicSlots } from '../../types/utils'

const appConfig = _appConfig as AppConfig & { b24ui: { descriptionList: Partial<typeof theme> } }

const descriptionList = tv({ extend: tv(theme), ...(appConfig.b24ui?.descriptionList || {}) })

type DescriptionListVariants = VariantProps<typeof descriptionList>

export interface DescriptionListItem {
  label?: string
  icon?: IconComponent
  avatar?: AvatarProps
  slot?: string
  description?: string
  /**
   * Display a list of actions:
   * - under the description if multiline
   * - next to the description if not multiline
   * `{ size: 'xs' }`{lang="ts-type"}
   */
  actions?: ButtonProps[]
  class?: any
  b24ui?: Partial<typeof descriptionList.slots>
}

export interface DescriptionListProps<T> {
  legend?: string
  text?: string
  /**
   * The key used to get the label from the item.
   * @defaultValue 'label'
   */
  labelKey?: string
  /**
   * The key used to get the description from the item.
   * @defaultValue 'description'
   */
  descriptionKey?: string
  items?: T[]
  size?: DescriptionListVariants['size']
  class?: any
  b24ui?: Partial<typeof descriptionList.slots>
}

type SlotProps<T> = (props: { item: T, index: number }) => any

export type DescriptionListSlots<T extends { slot?: string }> = {
  legend(props?: {}): any
  text(props?: {}): any
  leading: SlotProps<T>
  label: SlotProps<T>
  description: SlotProps<T>
  actions: SlotProps<T>
  footer(props?: { b24ui: any }): any
} & DynamicSlots<T, SlotProps<T>>
</script>

<script setup lang="ts"  generic="T extends DescriptionListItem">
import { computed } from 'vue'
import { get } from '../../utils'
import B24Avatar from '../Avatar.vue'
import B24Button from '../Button.vue'

const props = withDefaults(defineProps<DescriptionListProps<T>>(), {
  labelKey: 'label',
  descriptionKey: 'description'
})
const slots = defineSlots<DescriptionListSlots<T>>()

const b24ui = computed(() => descriptionList({
  size: props.size
}))

function normalizeItem(item: any) {
  const label = get(item, props.labelKey as string)
  const description = get(item, props.descriptionKey as string)
  const multiline = (item.actions || []).length > 1

  return {
    ...item,
    label,
    description,
    multiline
  }
}

const normalizedItems = computed(() => {
  if (!props.items) {
    return []
  }

  return props.items.map(normalizeItem)
})
</script>

<template>
  <div
    :class="b24ui.root({ class: [props.class, props.b24ui?.root] })"
  >
    <h2 v-if="legend || !!slots.legend" :class="b24ui.legend({ class: props.b24ui?.legend })">
      <slot name="legend">
        {{ legend }}
      </slot>
    </h2>
    <p v-if="text || !!slots.text" :class="b24ui.text({ class: props.b24ui?.text })">
      <slot name="text">
        {{ text }}
      </slot>
    </p>

    <dl :class="b24ui.container({ class: props.b24ui?.container })">
      <template
        v-for="(item, index) in normalizedItems"
        :key="index"
      >
        <dt
          :class="b24ui.labelWrapper({
            class: [
              props.b24ui?.labelWrapper,
              item?.b24ui?.labelWrapper
            ]
          })"
        >
          <slot name="leading" :item="item" :index="index">
            <Component
              :is="item.icon"
              v-if="item.icon"
              :class="b24ui.icon({
                class: [
                  props.b24ui?.icon,
                  item?.b24ui?.icon
                ]
              })"
            />
            <B24Avatar
              v-else-if="item.avatar"
              :size="((props.b24ui?.avatarSize || b24ui.avatarSize()) as AvatarProps['size'])"
              v-bind="item.avatar"
              :class="b24ui.avatar({
                class: [
                  props.b24ui?.avatar,
                  item?.b24ui?.avatar
                ]
              })"
            />
          </slot>
          <span
            :class="b24ui.label({
              class: [
                item?.class,
                props.b24ui?.label,
                item?.b24ui?.label
              ]
            })"
          >
            <slot name="label" :item="item" :index="index">
              {{ item.label }}
            </slot>
          </span>
        </dt>
        <dd
          :class="b24ui.descriptionWrapper({
            class: [
              props.b24ui?.descriptionWrapper,
              item?.b24ui?.descriptionWrapper
            ],
            multiline: item.multiline
          })"
        >
          <span
            :class="b24ui.description({
              class: [
                item?.class,
                props.b24ui?.description,
                item?.b24ui?.description
              ],
              multiline: item.multiline
            })"
          >
            <slot name="description" :item="item" :index="index">
              {{ item.description }}
            </slot>
          </span>
          <span
            v-if="item.actions?.length || !!slots.actions"
            :class="b24ui.actions({
              class: [
                props.b24ui?.actions,
                item?.b24ui?.actions
              ],
              multiline: item.multiline
            })"
          >
            <slot name="actions" :item="item" :index="index">
              <B24Button
                v-for="(action, indexActions) in item.actions"
                :key="indexActions"
                size="xs"
                v-bind="action"
              />
            </slot>
          </span>
        </dd>
      </template>
    </dl>
    <div
      v-if="!!slots.footer"
      :class="b24ui.footer({ class: props.b24ui?.footer })"
    >
      <slot name="footer" :b24ui="b24ui" />
    </div>
  </div>
</template>
