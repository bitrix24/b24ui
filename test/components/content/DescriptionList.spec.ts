import { describe, it, expect } from 'vitest'
// import DescriptionList, { type DescriptionListProps, type DescriptionListSlots } from '../../../src/runtime/components/content/DescriptionList.vue'
import DescriptionList, { type DescriptionListProps } from '../../../src/runtime/components/content/DescriptionList.vue'
import ComponentRender from '../../component-render'

describe('DescriptionList', () => {
  it.each([
    // Props
    ['with as', { props: { as: 'section' } }],
    ['with class', { props: { class: '' } }],
    ['with b24ui', { props: { b24ui: {} } }],
    // Slots
    ['with default slot', { slots: { default: () => 'Default slot' } }]
  // ])('renders %s correctly', async (nameOrHtml: string, options: { props?: DescriptionListProps, slots?: Partial<DescriptionListSlots> }) => {
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: DescriptionListProps }) => {
    const html = await ComponentRender(nameOrHtml, options, DescriptionList)
    expect(html).toMatchSnapshot()
  })
})
