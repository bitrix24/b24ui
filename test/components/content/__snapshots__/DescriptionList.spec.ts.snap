// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DescriptionList > renders with as correctly 1`] = `
"<div class="w-full shrink-0" as="section">
  <h2 class="text-base/7 font-semibold text-base-950 sm:text-sm/6 dark:text-white"> Order #1011 </h2>
  <p class="mt-0.5 max-w-2/3 text-sm/6 text-gray-500"> Personal details and application. </p>
  <dl class="mt-4 grid grid-cols-1 text-base/6 sm:grid-cols-[min(50%,theme(spacing.80))_auto] sm:text-sm/6">
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Customer </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Leslie Alexander </dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Email </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"><a href="mailto:leslie.alexander@example.com">leslie.alexander@example.com</a></dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Access </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Admin </dd>
  </dl>
</div>"
`;

exports[`DescriptionList > renders with b24ui correctly 1`] = `
"<div class="w-full shrink-0" b24ui="[object Object]">
  <h2 class="text-base/7 font-semibold text-base-950 sm:text-sm/6 dark:text-white"> Order #1011 </h2>
  <p class="mt-0.5 max-w-2/3 text-sm/6 text-gray-500"> Personal details and application. </p>
  <dl class="mt-4 grid grid-cols-1 text-base/6 sm:grid-cols-[min(50%,theme(spacing.80))_auto] sm:text-sm/6">
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Customer </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Leslie Alexander </dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Email </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"><a href="mailto:leslie.alexander@example.com">leslie.alexander@example.com</a></dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Access </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Admin </dd>
  </dl>
</div>"
`;

exports[`DescriptionList > renders with class correctly 1`] = `
"<div class="w-full shrink-0">
  <h2 class="text-base/7 font-semibold text-base-950 sm:text-sm/6 dark:text-white"> Order #1011 </h2>
  <p class="mt-0.5 max-w-2/3 text-sm/6 text-gray-500"> Personal details and application. </p>
  <dl class="mt-4 grid grid-cols-1 text-base/6 sm:grid-cols-[min(50%,theme(spacing.80))_auto] sm:text-sm/6">
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Customer </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Leslie Alexander </dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Email </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"><a href="mailto:leslie.alexander@example.com">leslie.alexander@example.com</a></dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Access </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Admin </dd>
  </dl>
</div>"
`;

exports[`DescriptionList > renders with default slot correctly 1`] = `
"<div class="w-full shrink-0">
  <h2 class="text-base/7 font-semibold text-base-950 sm:text-sm/6 dark:text-white"> Order #1011 </h2>
  <p class="mt-0.5 max-w-2/3 text-sm/6 text-gray-500"> Personal details and application. </p>
  <dl class="mt-4 grid grid-cols-1 text-base/6 sm:grid-cols-[min(50%,theme(spacing.80))_auto] sm:text-sm/6">
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Customer </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Leslie Alexander </dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Email </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"><a href="mailto:leslie.alexander@example.com">leslie.alexander@example.com</a></dd>
    <dt class="col-start-1 border-t border-base-950/5 pt-3 text-base-500 sm:border-t sm:border-base-950/5 sm:py-3 dark:border-white/5 dark:text-base-400 sm:dark:border-white/5"> Access </dt>
    <dd class="pb-3 pt-1 text-base-950 sm:border-t sm:border-base-950/5 sm:py-3 dark:text-white dark:sm:border-white/5"> Admin </dd>
  </dl>
</div>"
`;
